### ISSUE TITLE
Jupyter Server Config

### ISSUE THREAD

**Issue by alaws-USGS** [hytest-org/hytest#34]
Gene:
Starting this issue as a container to hold discussion and options about spinning up the jupyter server on the HPC host.

Here's the deal.... the Pangeo documentation -- and the how-to guidance written by rsignell (which is based on the pangeo docs) -- directs us to do the following to spin up the server:

on the HPC:
create a compute node: salloc -A username .....
Start the server itself: jupyter lab --no-browser .....
On your local desktop:
Set up a port-forwarder: ssh -N -L .....
Launch your web browser with http://localhost:8889
The jupyter server runs on the HPC host, and accepts connections on a given port number. Connections to that port using HTTP protocol will connect to the jupyter compute server.

I question the need to use a port-forwarder on my localhost. The way I read the docs, that step is used to forward connections made to my desktop on port 8889 toward the jupter server port on the HPC. Why can't I just connect directly to the HPC port from my browser?

The jupyter sever documentation describes this in some detail.... I think it would be a lot easier for the users if we could eliminate the ssh step from the flow.

This ssh step may be necessary because of network security going to and from the HPC host, but I will need to get access to the host in order to test whether that is actually true.

The set-up I am driving toward would be slightly more complicated to initially configure (but very much suited to automation). The advantage is that the repeated workflows (spin up server, run notebooks) would be greatly simplified.

The server configuration can be set in a few ways:

 command line switches
 ~/.jupyter/jupyter_server_config.py
 ~/.jupyter/jupyter_server_config.json
Command line is a nightmare for users... to avoid. I prefer editing the JSON file, as it is the easiest to understand and it only contains data for overriding defaults. The .py file is lengthy and more complex than we need.

Recommendations:

Generate a password (this automatically gets stored in jupyter_server_config.json
Add to the JSON config with these options:
Allow remote access
Bind to all interfaces (not just the one mapping to hostname).
Specify a certificate and key file to enable SSL. Our cert will be unregistered/serf-signed, so the browser will balk... but if we want https, we have to have a cert.
Set open_browser to False
Formally set the binding port per user account.
If we did all that, the user could spin up their jupyter server with a simple command: jupyter lab (rather than jupyter lab --no-browser --ip='hostname' --port=8400)

AND... we would eliminate all possibility of a port number collision among users. Firmly fixing the port number will also help with the friction of a self-signed certificate for SSL.

A few more details....  The fully-configured JSON that I have had good luck with (so far) is as follows: 
```json
{
  "ServerApp": {
    "password": << hashed password >> ,
    "open_browser": false, 
    "certfile" : "/Users/trantham/.jupyter/CERT.pem",
    "keyfile" : "/Users/trantham/.jupyter/KEY.key",
    "ip": "0.0.0.0",
    "port" : 8402,
    "allow_remote_access": true
  }
}
```
The hashed password is created a few different ways... We can call `jupyter server password` on the command line (this is what is shown in Pangeo docs).... or we can do it programatically within python: 
```python
from notebook.auth import passwd
hashed_pw = passwd()
```
That call to `passwd()` will read from the **terminal** (not STDIN) and generate a suitable hash of the password to store in its encrypted form.

The reason I like the idea of doing it from within a Python script is that we will be doing other things with python at setup time (like re-writing that JSON). IMO, it would be best to do it all in the same programming environment -- with a slick setup script. 

We also need to decide a couple of things on SSL and https. . . 

* Do we think it is necessary?  Since we're on a secured network, is it worth the trouble to set up SSL to encrypt the back-and-forth between the client and server? One could argue that it really is not required.  To-be-discussed @alaws-USGS 
* If we think that it is worth-while, so we want to 'publish' the cert file within the repo (causing every user to have the same self-signed cert), or have each potential user create their own certificate at install time?  For-sure, the better option is to have each user create their own cert... but it adds complexity to the setup....  See the example: 

```sh
> openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout KEY.key -out CERT.pm  
Generating a RSA private key
.....................................................................+++++
.......................................................................................................................................................................................................+++++
writing new private key to 'KEY.key'
-----
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [AU]: US
State or Province Name (full name) [Some-State]: Colorado
Locality Name (eg, city) []: Denver
Organization Name (eg, company) [Internet Widgits Pty Ltd]: DOI
Organizational Unit Name (eg, section) []: USGS 
Common Name (e.g. server FQDN or YOUR name) []: HyTEST
Email Address []:
```
That cert is good for a year, after which it will need to be re-generated. 

I found an easy way to automate everything... user just needs to supply a password.

See the file auto-conf.py

We still need to decide if we care about https and SSL. But if we do, we can automate every part of setting it up.

**Comment by alaws-USGS** [comment]
Andrew:
@gzt5142 I read through the comments and have a couple thoughts.

To the "We also need to decide a couple of things on SSL and https. . ." comment: You are correct that much of the work on HPCs will be over a secured network (VPN) but I think after our meeting today it would be a good idea to add SSL for any cloud or other use cases. And as for .pem/.crt files, would any of the default ones we use for DOI certs be useful? Info at https://github.com/usgs/best-practices/blob/master/ssl/UsingGitWithHTTPSInterception.md.

Rich:
The USGS-specific instructions we have for getting started on HPC are here:
https://github.com/USGS-python/hytest-development/blob/main/hpc.md

From Linux, Mac & Windowss/WSL2, we can also use this approach:
https://jupyter-forward.readthedocs.io/en/stable/

Andrew: Thanks Rich!

Gene:
I had not seen jupyter-forward before -- I had been focused instead on the hpc.md run-book.

It looks to me like jupyter-forward might be a more comprehensive solution, and also more flexible -- not requiring user to start an interactive shell on the HPC.

I wonder if we should pick one as the One True Solution(tm), and remove references to alternates. that is... either (1) coach users how to set things up from an HPC shell/login for execution from same.... or (2) use jupyter-forward from their desktop. Is there an advantage to having both? My take would be that simpler is better, so users can focus on the science they want to do.

If we present jupyter-forward as the best option, then we need to help users set up a conda env on their windows/mac/etc host and get jupyter-forward into their path. And then we need to offer a way to get the HyTEST-specific conda environment set up on HPC so that future runs of jupyter-forward will be able to activate it.

Rich:
@gzt5142 , I like the jupyter-forward approach a lot -- the main stumbling block was that we have many users on Windows, but there is an issue with port forwarding on windows. ( If you have any windows experience/expertise, it would be great if you could take a look at that issue!)

So currently Windows users would need to install WSL2 to use jupyter-forward -- which is great if they already have it, but if not, it's a bit of a high-bar or tough love situation. This is why we still kind of have the old instructions -- at least they work for everyone. :(

That said, if you do have WSL2 installed, it works nicely: ncar-xdev/jupyter-forward#152 (comment)

**Comment by alaws-USGS** [comment]
Gene:

Sad to say that I'm a relative newbie to the windows world, so not of immediate help. 

I think I still need to do some background research to understand the need for port forwarding. I'm still living in the naive world where I imagine an addressable interface when the jupyter server spins up. 

For example... the following works fine for me without setting up a port forwarder

On the remote host:
```text
jupyter lab --no-browser --port=8080 --ip='0.0.0.0'
```

Then in my desktop's browser:
```text
http://192.168.60.4:8080/lab
```
_(I have to specify the IP, since I am using a temporary virtual host for which there is no name in the DNS)._

The `--ip=0.0.0.0` replaces an older jupyter option for `allow_remote_access`. It instructs jupyter server to listen on all interfaces on the host (not just _localhost_).

Now... there may very well be a good reason this won't work on the HPC.  If I ever get access, I can test and try. 

The other thing to consider is that with `ssh`, the port forwarding takes place on a secured (SSL) connection. If we fell that is important, we can get jupyer server to do that also.  

**IF** that specification works on the HPC, this would eliminate the need for port forwarding on a windows host, and we could consolidate around jupyter-forward. 

So.... jupyter-forward uses ssh to execute the jupyter lab command on the remote (HPC) host. 

And it explicitly names the `--ip=<hostname>` parameter.  the host name comes from executing `hostname -f` on the remote host.  My understanding is that if that name resolves via DNS, then no port forwarding is needed to connect to the server.  (I am unable to test this at present).  The *wildcard* approach/solution such that port forwarding is not necessary would be to override the hostname so that it specifies `--ip=0.0.0.0` instead. 

There does not seem to be an easy way to do this without modifying the jupyter-forward code itself.  Using a custom version of this tool seems as difficult and 'tough love' an operation as getting folks to set up WSL2 on their Windows hosts.

If this solution seems worthwhile, we can pursue it further.... 

An ugly change just to test this.... changed line 189 of jupyter_forward/core.py to:
```
command = r'jupyter lab --no-browser --ip=0.0.0.0'
```

Then this command works such that no port-forwarding is needed: 

```
./jupyter-forward --identity ./.vagrant/machines/default/virtualbox/private_key --no-port-forwarding vagrant@127.0.0.1:2222
```

Rich:
Just a note that I tested these commands and they both worked (requesting that a 2 hour session be launched that uses the "woodshole" account and the "pangeo" conda environment).
```
$ jupyter-forward tallgrass --launch-command "srun -A woodshole -N 1 -t 02:00:00" --conda-env pangeo --port 8889

$ jupyter-forward denali --launch-command "srun -A woodshole -N 1 -t 02:00:00" --conda-env pangeo --port 8889
```
This is taking advantage that I have a `~/.ssh/config` file that contains these lines:
```
Host denali
     HostName denali.cr.usgs.gov
     User rsignell
     IdentityFile ~/.ssh/id_ed25519
Host tallgrass
     HostName tallgrass.cr.usgs.gov
     User rsignell
     IdentityFile ~/.ssh/id_ed25519
```

**Comment by gzt5142** [comment]
seems like this is a parallel issue to #7 ... should probably close this an put future discussion there.

**Comment by rsignell-usgs** [comment]
@gzt5142 and @alaws-USGS , great news - Jupyter-Forward is now working on Windows!  No WSL2 required!
https://github.com/ncar-xdev/jupyter-forward/issues/81#issuecomment-1242137881

**Comment by gzt5142** [comment]
SWEET.  I'll drive forward with this as the preferred option for running on HPC.

**Comment by gzt5142** [comment]
New information....   

Have been pointed toward OnDemand service at https://tallgrass-denali-ondemand.cr.usgs.gov/pun/sys/dashboard

At first blush, this looks like a much more streamlined user experience. 
User can request an allocation on tallgrass, within which a jupyter server will run.  User can custom-configure a conda environment.  

Am testing now to see how this works vs jupyter-forward or a manual HPC config.

**Comment by gzt5142** [comment]
Instructions for HPC launch using jupyter server are included in the PR #88 .
